File: aio/utils/dates.py
"""Natural language date parsing and formatting."""

from datetime import date, datetime, timedelta
from typing import TYPE_CHECKING

import dateparser

from aio.exceptions import InvalidDateError

if TYPE_CHECKING:
    pass


def parse_date(date_str: str) -> date:
    """Parse a natural language date string.

    Supports formats like:
    - "tomorrow", "today", "yesterday"
    - "next monday", "friday"
    - "in 3 days", "in a week"
    - "end of week", "end of month"
    - ISO format: "2024-01-15"

    Args:
        date_str: A natural language date string.

    Returns:
        A date object.

    Raises:
        InvalidDateError: If the date string cannot be parsed.
    """
    if not date_str or not date_str.strip():
        raise InvalidDateError("Date string cannot be empty")

    # Try parsing with dateparser
    result = dateparser.parse(
        date_str,
        settings={
            "PREFER_DATES_FROM": "future",
            "PREFER_DAY_OF_MONTH": "first",
            "RETURN_AS_TIMEZONE_AWARE": False,
        },
    )

    if result is None:
        raise InvalidDateError(f"Could not parse date: {date_str}")

    return result.date()


def format_relative_date(d: date) -> str:
    """Format a date as a relative string for display.

    Args:
        d: The date to format.

    Returns:
        A human-readable relative date string.
    """
    today = date.today()
    delta = (d - today).days

    if delta < -1:
        return f"{abs(delta)} days ago"
    elif delta == -1:
        return "yesterday"
    elif delta == 0:
        return "today"
    elif delta == 1:
        return "tomorrow"
    elif delta < 7:
        return d.strftime("%A")  # Day name
    elif delta < 14:
        return "next " + d.strftime("%A")
    else:
        return d.strftime("%b %d")  # "Jan 15"


def is_overdue(d: date) -> bool:
    """Check if a date is in the past.

    Args:
        d: The date to check.

    Returns:
        True if the date is before today.
    """
    return d < date.today()


def is_due_today(d: date) -> bool:
    """Check if a date is today.

    Args:
        d: The date to check.

    Returns:
        True if the date is today.
    """
    return d == date.today()


def is_due_this_week(d: date) -> bool:
    """Check if a date is within the next 7 days.

    Args:
        d: The date to check.

    Returns:
        True if the date is within the next 7 days (including today).
    """
    today = date.today()
    return today <= d <= today + timedelta(days=7)


def format_iso_date(d: date) -> str:
    """Format a date as ISO 8601.

    Args:
        d: The date to format.

    Returns:
        ISO 8601 formatted date string (YYYY-MM-DD).
    """
    return d.isoformat()


def format_iso_datetime(dt: datetime) -> str:
    """Format a datetime as ISO 8601.

    Args:
        dt: The datetime to format.

    Returns:
        ISO 8601 formatted datetime string.
    """
    return dt.isoformat()

File: obsidian-aio/src/utils/dates.ts
export class InvalidDateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InvalidDateError';
  }
}

/**
 * Parse a natural language date string.
 * Mirrors Python's aio/utils/dates.py parse_date()
 *
 * Supports formats like:
 * - "tomorrow", "today", "yesterday"
 * - "next monday", "friday"
 * - "in 3 days", "in a week"
 * - "end of week", "end of month", "end of year"
 * - ISO format: "2024-01-15"
 */
export function parseDate(dateStr: string): Date {
  if (!dateStr?.trim()) {
    throw new InvalidDateError('Date string cannot be empty');
  }

  const today = new Date();
  const lower = dateStr.toLowerCase().trim();

  // ISO format: YYYY-MM-DD
  if (/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
    const parsed = new Date(dateStr + 'T00:00:00');
    if (isNaN(parsed.getTime())) {
      throw new InvalidDateError(`Invalid date: ${dateStr}`);
    }
    return parsed;
  }

  // Today
  if (lower === 'today') {
    return today;
  }

  // Tomorrow
  if (lower === 'tomorrow') {
    const tomorrow = new Date(today);
    tomorrow.setDate(tomorrow.getDate() + 1);
    return tomorrow;
  }

  // Yesterday
  if (lower === 'yesterday') {
    const yesterday = new Date(today);
    yesterday.setDate(yesterday.getDate() - 1);
    return yesterday;
  }

  // Next week
  if (lower === 'next week') {
    const nextWeek = new Date(today);
    nextWeek.setDate(nextWeek.getDate() + 7);
    return nextWeek;
  }

  // "in X days"
  const inDaysMatch = lower.match(/^in (\d+) days?$/);
  if (inDaysMatch) {
    const days = parseInt(inDaysMatch[1], 10);
    const future = new Date(today);
    future.setDate(future.getDate() + days);
    return future;
  }

  // "in X weeks"
  const inWeeksMatch = lower.match(/^in (\d+) weeks?$/);
  if (inWeeksMatch) {
    const weeks = parseInt(inWeeksMatch[1], 10);
    const future = new Date(today);
    future.setDate(future.getDate() + weeks * 7);
    return future;
  }

  // Day names: "monday", "next monday", etc.
  const days = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];
  const nextDayMatch = lower.match(/^(next )?(monday|tuesday|wednesday|thursday|friday|saturday|sunday)$/);
  if (nextDayMatch) {
    const targetDay = days.indexOf(nextDayMatch[2]);
    const currentDay = today.getDay();
    let daysAhead = targetDay - currentDay;
    if (daysAhead <= 0 || nextDayMatch[1]) {
      daysAhead += 7;
    }
    const targetDate = new Date(today);
    targetDate.setDate(targetDate.getDate() + daysAhead);
    return targetDate;
  }

  // End of week / eow (Friday)
  if (lower === 'end of week' || lower === 'eow') {
    return getEndOfWeek();
  }

  // End of month / eom
  if (lower === 'end of month' || lower === 'eom') {
    return getEndOfMonth();
  }

  // End of year / eoy
  if (lower === 'end of year' || lower === 'eoy') {
    return getEndOfYear();
  }

  throw new InvalidDateError(`Could not parse date: ${dateStr}`);
}

/**
 * Format a date as a relative string for display.
 * Mirrors Python's format_relative_date()
 */
export function formatRelativeDate(d: Date): string {
  const today = startOfDay(new Date());
  const target = startOfDay(d);
  const delta = Math.round(
    (target.getTime() - today.getTime()) / (1000 * 60 * 60 * 24)
  );

  if (delta < -1) return `${Math.abs(delta)} days ago`;
  if (delta === -1) return 'yesterday';
  if (delta === 0) return 'today';
  if (delta === 1) return 'tomorrow';
  if (delta < 7) return d.toLocaleDateString('en-US', { weekday: 'long' });
  if (delta < 14)
    return 'next ' + d.toLocaleDateString('en-US', { weekday: 'long' });
  return d.toLocaleDateString('en-US', { month: 'short', day: 'numeric' });
}

/**
 * Check if a date is in the past.
 */
export function isOverdue(d: Date): boolean {
  return startOfDay(d) < startOfDay(new Date());
}

/**
 * Check if a date is today.
 */
export function isDueToday(d: Date): boolean {
  return startOfDay(d).getTime() === startOfDay(new Date()).getTime();
}

/**
 * Check if a date is within the next 7 days (including today).
 */
export function isDueThisWeek(d: Date): boolean {
  const today = startOfDay(new Date());
  const target = startOfDay(d);
  const weekFromNow = new Date(today);
  weekFromNow.setDate(weekFromNow.getDate() + 7);
  return target >= today && target <= weekFromNow;
}

/**
 * Format a date as ISO 8601 (YYYY-MM-DD).
 */
export function formatIsoDate(d: Date): string {
  return d.toISOString().split('T')[0];
}

// Helper: Get start of day (midnight)
function startOfDay(d: Date): Date {
  const result = new Date(d);
  result.setHours(0, 0, 0, 0);
  return result;
}

// Helper: Get next Friday (or Friday next week if today is Friday-Sunday)
function getEndOfWeek(): Date {
  const today = new Date();
  const dayOfWeek = today.getDay();
  let daysUntilFriday = 5 - dayOfWeek;
  if (daysUntilFriday <= 0) {
    daysUntilFriday += 7;
  }
  const friday = new Date(today);
  friday.setDate(today.getDate() + daysUntilFriday);
  return friday;
}

// Helper: Get last day of current month
function getEndOfMonth(): Date {
  const today = new Date();
  return new Date(today.getFullYear(), today.getMonth() + 1, 0);
}

// Helper: Get December 31 of current year
function getEndOfYear(): Date {
  const today = new Date();
  return new Date(today.getFullYear(), 11, 31);
}
